### ISSUE TITLE
MDB_NOTFOUND stacktrace in the logs while working on txhashet

### ISSUE THREAD

**Issue by mably** [mimblewimble/grin#3250]
**Describe the bug**
No idea what the bug is.  If there is a bug.

**To Reproduce**
No idea.  Just seen it a few times in the logs.

**Relevant Information**
```20200227 14:33:59.282 DEBUG grin_p2p::protocol - handle_payload: txhashset req for 0000600ce353 at 582092
20200227 14:33:59.282 DEBUG grin_chain::chain - txhashset_archive_header: body_head - 00000a9701ce, 584972, txhashset height - 581760
20200227 14:33:59.282 DEBUG grin_chain::txhashset::txhashset - Rewind header extension to 00005f0a9f7c at 581760 from 00000a9701ce at 584972
20200227 14:33:59.401 DEBUG grin_chain::txhashset::txhashset - Rewind extension to 00005f0a9f7c at 581760 from 00000a9701ce at 584972
20200227 14:33:59.810 WARN grin_chain::txhashset::txhashset - rewind_single_block: fallback to legacy input bitmap for block 00005bfbca4e at 582084
20200227 14:33:59.811 WARN grin_chain::txhashset::txhashset - rewind_single_block: fallback to legacy input bitmap for block 000036b8f92d at 582083
20200227 14:33:59.811 WARN grin_chain::txhashset::txhashset - rewind_single_block: fallback to legacy input bitmap for block 0000091b005a at 582082
20200227 14:33:59.811 WARN grin_chain::txhashset::txhashset - rewind_single_block: fallback to legacy input bitmap for block 00005bec8da0 at 582081
20200227 14:33:59.812 WARN grin_chain::txhashset::txhashset - rewind_single_block: fallback to legacy input bitmap for block 0000078f25d4 at 582080
20200227 14:33:59.814 WARN grin_servers::common::adapters - Couldn't produce txhashset data for block 00005f0a9f7c: Error { inner: LmdbErr(Error::Code(-30798, 'MDB_NOTFOUND: No matching key/data pair found'))

stack backtrace:
   0: failure::backtrace::internal::InternalBacktrace::new
   1: <failure::backtrace::Backtrace as core::default::Default>::default
   2: failure::context::Context<D>::with_err
   3: <grin_chain::error::Error as core::convert::From<grin_store::lmdb::Error>>::from
   4: grin_chain::txhashset::txhashset::Extension::rewind
   5: grin_chain::pipe::rewind_and_apply_fork
   6: grin_chain::txhashset::txhashset::extending_readonly
   7: grin_chain::chain::Chain::txhashset_read
   8: <grin_servers::common::adapters::NetToChainAdapter as grin_p2p::types::ChainAdapter>::txhashset_read
   9: <grin_p2p::peers::Peers as grin_p2p::types::ChainAdapter>::txhashset_read
  10: <grin_p2p::peer::TrackingAdapter as grin_p2p::types::ChainAdapter>::txhashset_read
  11: <grin_p2p::protocol::Protocol as grin_p2p::conn::MessageHandler>::consume
  12: std::sys_common::backtrace::__rust_begin_short_backtrace
  13: std::panicking::try::do_call
  14: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  15: core::ops::function::FnOnce::call_once{{vtable.shim}}
  16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
  17: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
      std::sys_common::thread::start_thread
             at src/libstd/sys_common/thread.rs:13
      std::sys::unix::thread::Thread::new::thread_start
             at src/libstd/sys/unix/thread.rs:80
  18: start_thread
  19: __clone


Store Error: LmdbErr(Error::Code(-30798, 'MDB_NOTFOUND: No matching key/data pair found')), reason: LMDB error }```

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Grin Version 3.1.0-beta.2

**Additional context**
N/A

**Comment by antiochp** [comment]
@mably thanks for the bug report. I have a vague theory of what is going on here. I need to spend some time digging into it a bit more.

Just to be clear - your node is still running ok even after seeing this error? Its not preventing your node from staying in sync with latest blocks?

**Comment by mably** [comment]
Yep everything seems normal.

Though I'm connected to a bit less peers than before : ~30 instead of ~80-90.

**Comment by antiochp** [comment]
One more question - Was this a node that sync'd recently or was this a node that has been receiving blocks during broadcast for a while?

**Comment by mably** [comment]
Nope, rather old node, didn't sync recently.

**Comment by antiochp** [comment]
So this was an existing `chain_data` and the code was upgraded to `3.1.0-beta.2` and then this error occurred?

**Comment by mably** [comment]
Yep exactly.

**Comment by mably** [comment]
Problem hasn't occured again since yesterday.  Could it be related to some outdated peer?

**Comment by antiochp** [comment]
This is happening when a peer requests a txhashset from your node.
Your node then rewinds to a known block height, makes a copy of the txhashset files and zips them up.
The issue is happening during this long rewind (we rewind back approx 2 days worth of history).

3.0.0 stored a bitmap of spent outputs per block.
3.1.0 stores a slightly different set of data for "spent" info per block.

The issue is actually happening when a rewind goes back beyond 3.1.0 data and back into "legacy" 3.0.0 data. So will only occur for at most a couple of days. It is likely your node can now safely rewind across a full set of data.

It looks like the error is happening because we cannot find a 3.0.0 bitmap in the db for a block being rewound.
But its actually misleading. Its not the bitmap that cannot be found, its actually during the subsequent cleanup of the output_pos index where we delete index entries based on this bitmap - 

```
20200228 11:04:13.890 DEBUG grin_chain::txhashset::txhashset - ***** about to delete output_pos_height for: Commitment(086dc1df4011d17d8c13876cd61996343736042747d202b47ccee4dcec2aa48bba)
20200228 11:04:13.891 DEBUG grin_chain::txhashset::txhashset - ***** Err(LmdbErr(Error::Code(-30798, 'MDB_NOTFOUND: No matching key/data pair found')))
```

tl;dr 3.1.0 keeps the index more closely aligned with the unspent output set, but in doing so we were making an incorrect assumption about the underlying index.
We do not want to fail like this if we try and delete an entry that is missing. We just want to continue with the rewind.

Fix coming shortly.

**Comment by antiochp** [comment]
Thanks for spotting this in your logs @mably.
Hard to track down due to it only happening for a relatively short period after upgrading to 3.1.0.

**Comment by mably** [comment]
Happy to be able to help.  Thanks for your great work @antiochp 👍
